vmm/src/api/dbus/mod.rs

```diff
@@ -0,0 +1,364 @@
+    pub service_name: String,
+    pub object_path: String,
+    pub system_bus: bool,
+}
+
+pub struct DBusApi {
+    api_notifier: EventFd,
+    api_sender: futures::lock::Mutex<Sender<ApiRequest>>,
+}
+
+fn api_error(error: impl std::fmt::Debug) -> fdo::Error {
+    fdo::Error::Failed(format!("{error:?}"))
+}
+
+// This method is intended to ensure that the DBusApi thread has enough time to
+// send a response to the VmmShutdown method call before it is terminated. If
+// this step is omitted, the thread may be terminated before it can send a
+// response, resulting in an error message stating that the message recipient
+// disconnected from the message bus without providing a reply.
+pub fn dbus_api_graceful_shutdown(ch: DBusApiShutdownChannels) {
+    let (send_shutdown, mut recv_done) = ch;
+
+    // send the shutdown signal and return
+    // if it errors out
+    if send_shutdown.send(()).is_err() {
+        return;
+    }
+
+    // loop until `recv_err` errors out
+    // or as long as the return value indicates
+    // "immediately stale" (None)
+    while let Ok(None) = recv_done.try_recv() {}
+}
+
+impl DBusApi {
+    pub fn new(api_notifier: EventFd, api_sender: Sender<ApiRequest>) -> Self {
+        Self {
+            api_notifier,
+            api_sender: futures::lock::Mutex::new(api_sender),
+        }
+    }
+
+    async fn clone_api_sender(&self) -> Sender<ApiRequest> {
+        // lock the async mutex, clone the `Sender` and then immediately
+        // drop the MutexGuard so that other tasks can clone the
+        // `Sender` as well
+        self.api_sender.lock().await.clone()
+    }
+
+    fn clone_api_notifier(&self) -> Result<EventFd> {
+        self.api_notifier
+            .try_clone()
+            .map_err(|err| fdo::Error::IOError(format!("{err:?}")))
+    }
+
+    async fn vm_action(&self, action: VmAction) -> Result<Optional<String>> {
+        let api_sender = self.clone_api_sender().await;
+        let api_notifier = self.clone_api_notifier()?;
+
+        let result = blocking::unblock(move || super::vm_action(api_notifier, api_sender, action))
+            .await
+            .map_err(api_error)?
+            // We're using `from_utf8_lossy` here to not deal with the
+            // error case of `from_utf8` as we know that `b.body` is valid JSON.
+            .map(|b| String::from_utf8_lossy(&b.body).to_string());
+
+        Ok(result.into())
+    }
+}
+
+// TODO: no integration tests, implement them
+#[dbus_interface(name = "org.cloudhypervisor.DBusApi1")]
+impl DBusApi {
+    async fn vmm_ping(&self) -> Result<String> {
+        let api_sender = self.clone_api_sender().await;
+        let api_notifier = self.clone_api_notifier()?;
+
+        let result = blocking::unblock(move || super::vmm_ping(api_notifier, api_sender))
+            .await
+            .map_err(api_error)?;
+        serde_json::to_string(&result).map_err(api_error)
+    }
+
+    // TODO: VMM terminates before we can send out anything to
```

Comment:
Have you got to the bottom of this?

Reply:
Yeah, this issue was resolved with the graceful shutdown commit. I had forgotten to delete this while editing the commit history to make the review process easier.